File: scripts/acronym.js
// Description
//   Generates help commands for Hubot.
// 
// Commands:
//   !!acro <acronym> - Finds an acronym.
// 

var cheerio = require("cheerio");

module.exports = function (robot) {
  robot.hear(/^!!acro (.+)/i, function (res) {
    robot.getAcronyms(res, res.match[1].trim().split(" "));
  });
  
  robot.getAcronyms = function (res, list) {
    var completed = 0;
    var responses = [];
    var limit = 5;

    var max = list.length;
    if (list.length > limit) {
      max = limit;
    }

    // Requested by WBO, do not remove unless you get his express permission
    if(list.length == 1) {
      if(list[0].toLowerCase() == ":horse:" || list[0] == "horse") {
        res.send(">:taco:");
        return;
      } else if(list[0].toLowerCase() == ":rachel:" || list[0] == "rachel") {
        res.send(">:older_woman:");
        return;
      }
    }

    // Chain all the HTTP requests so its nice and synchronous
    for (var i = 0; i < max; i++) {
 
      // Keep everything in a closure so we can access the index variable later on
      (function (index) {
        robot.http("http://acronyms.thefreedictionary.com/" + list[index]).get()
        (function (err, resp, body) {
          if (!err) {
            responses[index] = body;
            completed++;
            
            // Called when completing the final request
            if (completed == max) {
              var response = "";
              for (var j = 0; j < responses.length; j++) {
                var acro = list[j].toUpperCase();
                var acronym = robot.getAcronym(responses[j]);
                response += ">" + acro + ": " + acronym + "\r\n";
              }
  
              if (list.length > 5) {
                response += ">I am limited to " + limit + " acronyms at once.";
              }
              res.send(response);
            }
          } else {
            return;
          }
        });
      })(i);
    }
  }
  
  // Parses the DOM to get the acroynm expansion
  robot.getAcronym = function (html) {
    var $ = cheerio.load(html);
    var responses = [];
    
    $('td.acr').each(function (i, element) {
      responses.push($(element).siblings().text());
    });
    
    return responses[Math.floor(Math.random() * responses.length)];
  }
}

File: uqcsbot/scripts/acronym.py
from uqcsbot import bot, Command
from requests import get
from requests.utils import quote
from bs4 import BeautifulSoup
from typing import List
from functools import partial
import asyncio

ACRONYM_LIMIT = 5
BASE_URL = "http://acronyms.thefreedictionary.com"


async def get_acronyms(loop, word: str) -> (str, List[str]):
    http_response = await loop.run_in_executor(None, partial(get, f"{BASE_URL}/{quote(word)}"))
    html = BeautifulSoup(http_response.content, 'html.parser')
    acronym_tds = html.find_all("td", class_="acr")
    return word, [td.find_next_sibling("td").text for td in acronym_tds]


@bot.on_command("acro")
def handle_acronym(command: Command):
    '''
    `!acro <TEXT>` - Finds an acronym for the given text.
    '''
    if not command.has_arg():
        return

    words = command.arg.split(" ")

    # Requested by @wbo, do not remove unless you get his express permission
    if len(words) == 1:
        word = words[0]
        if word.lower() in [":horse:", "horse"]:
            bot.post_message(command.channel, ">:taco:")
            return
        elif word.lower() in [":rachel:", "rachel"]:
            bot.post_message(command.channel, ">:older_woman:")
            return

    loop = bot.get_event_loop()
    acronym_futures = [get_acronyms(loop, word) for word in words[:ACRONYM_LIMIT]]
    response = ""
    for word, acronyms in loop.run_until_complete(asyncio.gather(*acronym_futures)):
        if acronyms:
            acronym = acronyms[0]
            response += f">{word.upper()}: {acronym}\r\n"
        else:
            response += f"{word.upper()}: No acronyms found!\r\n"

    if len(words) > ACRONYM_LIMIT:
        response += f">I am limited to {ACRONYM_LIMIT} acronyms at once"

    bot.post_message(command.channel, response)
